File: src/env.ts
import * as Exp from "./exp"
import * as Value from "./value"
import * as Err from "./err"
import { evaluate } from "./evaluate"

export class Env {
  constructor(public entry_map: Map<string, Entry.Entry> = new Map()) {}

  lookup_type_and_value(
    name: string
  ): undefined | { t: Value.Value; value: Value.Value } {
    let entry = this.entry_map.get(name)

    if (entry === undefined) {
      return undefined
    } else if (entry instanceof Entry.DefineRec) {
      let { t, value, env } = entry
      return {
        t: evaluate(env.ext_rec(name, { t, value, env }), t),
        value: evaluate(env.ext_rec(name, { t, value, env }), value),
      }
    } else if (entry instanceof Entry.Define) {
      let { t, value } = entry
      return { t, value }
    } else {
      throw new Err.Unhandled(entry)
    }
  }

  lookup_type(name: string): undefined | Value.Value {
    let result = this.lookup_type_and_value(name)

    if (result === undefined) {
      return undefined
    } else {
      let { t } = result
      return t
    }
  }

  lookup_value(name: string): undefined | Value.Value {
    let result = this.lookup_type_and_value(name)

    if (result === undefined) {
      return undefined
    } else {
      let { value } = result
      return value
    }
  }

  ext(
    name: string,
    the: {
      t: Value.Value
      value: Value.Value
    }
  ): Env {
    return new Env(
      new Map([...this.entry_map, [name, new Entry.Define(the.t, the.value)]])
    )
  }

  ext_rec(
    name: string,
    the: {
      t: Exp.Exp
      value: Exp.Exp
      env: Env
    }
  ): Env {
    return new Env(
      new Map([
        ...this.entry_map,
        [name, new Entry.DefineRec(the.t, the.value, the.env)],
      ])
    )
  }
}

export namespace Entry {
  export abstract class Entry {
    abstract_class_name: "Env.Entry" = "Env.Entry"
  }

  export class DefineRec extends Entry {
    constructor(public t: Exp.Exp, public value: Exp.Exp, public env: Env) {
      super()
    }
  }

  export class Define extends Entry {
    constructor(public t: Value.Value, public value: Value.Value) {
      super()
    }
  }
}

File: scala/src/main/scala/xieyuheng/cicada/Env.scala
package xieyuheng.cicada

import collection.immutable.ListMap

import evaluate._

sealed trait EnvEntry
final case class EnvEntryDefineRec(t: Exp, value: Exp, env: Env) extends EnvEntry
final case class EnvEntryDefine(t: Value, value: Value) extends EnvEntry

case class Env(entry_map: ListMap[String, EnvEntry] = ListMap()) {

  def lookup_type_and_value(name: String): Option[(Value, Value)] = {
    entry_map.get(name).map {
      case EnvEntryDefineRec(t: Exp, value: Exp, env: Env) =>
        (evaluate(env.ext_rec(name, t, value, env), t),
          evaluate(env.ext_rec(name, t, value, env), value))
      case EnvEntryDefine(t: Value, value: Value) =>
        (t, value)
    }
  }

  def lookup_type(name: String): Option[Value] = {
    lookup_type_and_value(name).map {
      case (t, _value) => t
    }
  }

  def lookup_value(name: String): Option[Value] = {
    lookup_type_and_value(name).map {
      case (_t, value) => value
    }
  }

  def ext(name: String, t: Value, value: Value): Env = {
    Env(this.entry_map + (name -> EnvEntryDefine(t, value)))
  }

  def ext_rec(name: String, t: Exp, value: Exp, env: Env): Env = {
    Env(this.entry_map + (name -> EnvEntryDefineRec(t, value, env)))
  }

}
